src/ng-select/items-list.ts

```diff
@@ -94,8 +101,30 @@ export class ItemsList {
     }
 
     filter(term: string) {
-        const filterFuncVal = this._getDefaultFilterFunc(term);
-        this._filteredItems = term ? this._items.filter(val => filterFuncVal(val)) : this._items;
+        if (!term) {
+            this._filteredItems = this._items;
+            return;
+        }
+
+        this._filteredItems = [];
+        term = searchHelper.stripSpecialChars(term).toUpperCase();
+        for (const key of Object.keys(this._groups)) {
```

Comment:
maybe this can be simplified firstly filtering on flat structure items that match term:
```javascript
this._filteredItems = this.items.filter(item => {
    const label = searchHelper.stripSpecialChars(item.label ? item.label.toString() : '').toLowerCase();
    return item.hasChildren || label.indexOf(term) > -1;
})
```
and then removing all groups that have no children?
Ideally it would be only two iterations

Reply:
I tried this version, but it is significant slower. In example with 5000 items there each item has one group(slowest possible scenario) I got result of 320ms for every filter call. Current implementation took only 13ms to complete. This is because you just need to search ones in each group and then only find head item and push it as first in group, also I'm not searching in group head items which is not needed.